components/equihash/tromp/equi_miner.c

```diff
@@ -0,0 +1,737 @@
+#elif WN == 200 && RESTBITS == 8
+      return (pslot->hash->bytes[htl->prevbo] & 0xf) << 4 | pslot->hash->bytes[htl->prevbo+1] >> 4;
+#elif WN == 200 && RESTBITS == 9
+      return (pslot->hash->bytes[htl->prevbo] & 0x1f) << 4 | pslot->hash->bytes[htl->prevbo+1] >> 4;
+#elif WN == 144 && RESTBITS == 4
+      return pslot->hash->bytes[htl->prevbo] & 0xf;
+#else
+#error non implemented
+#endif
+    }
+    u32 getxhash1(const htlayout *htl, const slot1* pslot) {
+#if WN == 200 && RESTBITS == 4
+      return pslot->hash->bytes[htl->prevbo] & 0xf;
+#elif WN == 200 && RESTBITS == 8
+      return pslot->hash->bytes[htl->prevbo];
+#elif WN == 200 && RESTBITS == 9
+      return (pslot->hash->bytes[htl->prevbo]&1) << 8 | pslot->hash->bytes[htl->prevbo+1];
+#elif WN == 144 && RESTBITS == 4
+      return pslot->hash->bytes[htl->prevbo] & 0xf;
+#else
+#error non implemented
+#endif
+    }
+    bool htlayout_equal(const htlayout *htl, const hashunit *hash0, const hashunit *hash1) {
+      return hash0[htl->prevhashunits-1].word == hash1[htl->prevhashunits-1].word;
+    }
+
+#if RESTBITS <= 6
+    typedef uchar xslot;
+#else
+    typedef u16 xslot;
+#endif
+  struct collisiondata {
+#ifdef XBITMAP
+#if NSLOTS > 64
+#error cant use XBITMAP with more than 64 slots
+#endif
+    u64 xhashmap[NRESTS];
+    u64 xmap;
+#else
+    xslot nxhashslots[NRESTS];
+    xslot xhashslots[NRESTS][XFULL];
+    xslot *xx;
+    u32 n0;
+    u32 n1;
+#endif
+    u32 s0;
+  };
+  typedef struct collisiondata collisiondata;
+
+    void collisiondata_clear(collisiondata *cd) {
+#ifdef XBITMAP
+      memset(cd->xhashmap, 0, NRESTS * sizeof(u64));
+#else
+      memset(cd->nxhashslots, 0, NRESTS * sizeof(xslot));
+#endif
+    }
+    bool addslot(collisiondata *cd, u32 s1, u32 xh) {
+#ifdef XBITMAP
+      xmap = xhashmap[xh];
+      xhashmap[xh] |= (u64)1 << s1;
+      s0 = -1;
+      return true;
+#else
+      cd->n1 = (u32)cd->nxhashslots[xh]++;
+      if (cd->n1 >= XFULL)
+        return false;
+      cd->xx = cd->xhashslots[xh];
+      cd->xx[cd->n1] = s1;
+      cd->n0 = 0;
+      return true;
+#endif
+    }
+    bool nextcollision(const collisiondata *cd) {
+#ifdef XBITMAP
+      return cd->xmap != 0;
+#else
+      return cd->n0 < cd->n1;
+#endif
+    }
+    u32 slot(collisiondata *cd) {
+#ifdef XBITMAP
+      const u32 ffs = __builtin_ffsll(cd->xmap);
+      s0 += ffs; cd->xmap >>= ffs;
+      return s0;
+#else
+      return (u32)cd->xx[cd->n0++];
+#endif
+    }
+
+  void equi_digit0(equi *eq, const u32 id) {
+    uchar hash[HASHOUT];
+    BLAKE2bState* state;
+    htlayout htl = htlayout_new(eq, 0);
+    const u32 hashbytes = hashsize(0);
+    for (u32 block = id; block < NBLOCKS; block++) {
+      state = eq->blake2b_clone(eq->blake_ctx);
+      u32 leb = htole32(block);
+      eq->blake2b_update(state, (uchar *)&leb, sizeof(u32));
+      eq->blake2b_finalize(state, hash, HASHOUT);
+      eq->blake2b_free(state);
+      // Avoid use-after-free and double-free
```

Comment:
It may be a good idea to add this warning to the `blake2b_free()` function docs too.